include/navigation/perfect_memory.h

```diff
@@ -115,29 +103,41 @@ class PerfectMemory : public VisualNavigationBase
         return {0, getNumSnapshots()};
     }
 
+    //! Get the resolution of images
+    const cv::Size &getUnwrapResolution() const { return m_UnwrapRes; }
+
 protected:
+    static const Window defaultWindow;
+
     //------------------------------------------------------------------------
     // Protected API
     //------------------------------------------------------------------------
     float calcSnapshotDifference(const cv::Mat &image,
                                  const ImgProc::Mask &mask, size_t snapshot) const
     {
-        return m_Store.calcSnapshotDifference(image, mask, snapshot, getMask());
+        return m_Store.calcSnapshotDifference(image, mask, snapshot);
     }
 
 private:
     //------------------------------------------------------------------------
     // Private members
     //------------------------------------------------------------------------
+    const cv::Size m_UnwrapRes;
     Store m_Store;
     mutable std::vector<float> m_Differences;
 
-    void testInternal(const Window &window, const cv::Mat &image) const
+    void testInternal(const cv::Mat &image, const ImgProc::Mask &mask, const Window &window) const
     {
         const auto &unwrapRes = getUnwrapResolution();
         BOB_ASSERT(image.cols == unwrapRes.width);
         BOB_ASSERT(image.rows == unwrapRes.height);
         BOB_ASSERT(image.type() == CV_8UC1);
+        if(!mask.empty()) {
+            BOB_ASSERT(mask.get().cols == unwrapRes.width);
+            BOB_ASSERT(mask.get().rows == unwrapRes.height);
+            BOB_ASSERT(mask.get().type() == CV_8UC1);
+        }
```

Comment:
You could replace this block with ``BOB_ASSERT(mask.isValid(unwrapRes));`` (which will also work if ``mask`` is empty).

You don't need to check the type as this will always be ``CV_8UC1`` as long as the API has been used properly.